Add additional actions to CommentThread

marrej opened the pull request:
This PR proposes extension defined additional actions for Comment Threads.  (See image below for reference).
![Comment](https://user-images.githubusercontent.com/13024177/194103735-ff5f6325-448c-4378-aa32-72a1e047087d.png)

To allow extensions to define additionalActions which are always visible in the CommentThread, a new area has been defined, right under the `commentReply`. This area contains separator & extension defined actions.
Extensions can defined both single commands and submenus.

Submenus will be rendered as Dropdown buttons, where the first action in the submenu will be treated as the primary action - due to the submenus not carrying Command.
If submenu contains only 1 action, then a single button will be created instead of dropdown button.
In case of no action attached to submenu, the submenu action would be displayed.

**NOTE:** if there are doubts about `AdditionalActions` being the correct name for this new area, we should change it.
____

- New file has been added to contain the functionality for the Additional Actions
  - Hides the area if there are no active actions or submenu actions
  - sends back only CommentThread to the extension 
- CommentThreadWidget invokes appends this new area in to the `CommentThreadBody` under the `CommentReply`
- CommentFormActions has been modified to provide dropdown buttons along the single buttons (Could be move to a separate file if deemed neccessary, although there would be some unnecessary code duplication)
- added styling for the new area & elements


partially fixes https://github.com/microsoft/vscode/issues/156885
fixes https://github.com/microsoft/vscode/issues/163281

alexr00 commented on src/vs/workbench/contrib/comments/browser/commentThreadAdditionalActions.ts:
```diff
@@ -0,0 +1,96 @@
+
+			action.run({
+				thread: this._commentThread,
+				$mid: MarshalledId.CommentThreadReply
```
If you want to use `MarshalledId.CommentThreadReply` here then there should also be a `text` property. Otherwise, it may make more sense to structure this as a `MarshalledId.CommentThread` and use that instead.

marrej replied:
I agree that CommentThread would make more sense, although that responds back only with a simplified reference to the CommentThread (ids) and not the actual thread object, so its not easily usable.
I would suggest that we add a CommentThreadInstance which would be able to retrieve the just the thread.
Otherwise we could fallback to the CommentThreadReply if you don't want to add the additional property to MarshalledIds.
What do you think?

alexr00 replied:
`CommentThreadInstance` works for me!

alexr00 reviewed (changes requested):
Thanks for the PR! This is looking good. I'm seeing some bad alignment of the actions row:

I also don't really like the separator, but I'd like to get @daviddossett's opinion on that.

marrej commented on src/vs/workbench/services/actions/common/menusExtensionPoint.ts:
```diff
@@ -160,6 +160,12 @@ const apiMenus: IAPIMenu[] = [
+		key: 'comments/commentThread/additionalActions',
+		id: MenuId.CommentThreadAdditionalActions,
+		description: localize('commentThread.actions', "The contributed comment thread context menu, rendered as buttons below the comment editor"),
+		supportsSubmenus: true
```
Makes sense, Added the contribution.

marrej commented on src/vs/workbench/contrib/comments/browser/commentThreadAdditionalActions.ts:
```diff
@@ -0,0 +1,98 @@
+
+		}, this.themeService, this.contextMenuService);
+
+		this._commentFormActions.setActions(menu);
```
Thanks for the catch! Added it in

marrej commented:
> Thanks for the PR! This is looking good. I'm seeing some bad alignment of the actions row

I tried to simply fix the problem with the bottom offsetting but it seems to me that there is a bigger problem in play & that is that the height calculation does not fullly align with the actual container height. This problem seems to occur in the actual version with replys as well

![image](https://user-images.githubusercontent.com/13024177/201788407-44a7ada2-97b4-4202-be2b-5f9c8cc4b7c2.png)
![image](https://user-images.githubusercontent.com/13024177/201788421-2ceaae3e-86d2-42e2-aefd-188f4ca8b816.png)
![image](https://user-images.githubusercontent.com/13024177/201788429-e6fb0e13-1386-483b-b5ba-d5dbf8690336.png)

So I would suggest we open a bug for that and treat it as a separate issue.
Although I have increase the spacing at the bottom of the AdditionActions so its not mushed against the bottom border.
- also changes the reply panel padding to margin so we don't get double offset between the additional actions & the reply area

> I also don't really like the separator, but I'd like to get @daviddossett's opinion on that.

The main reasoning for this from my point of view, is the clear separation between the actions connected to the Reply area & the CommentThread. Once the reply area is opened and reply buttons are shown, without the clear visual separation it may be a bit confusing for the users.

![image](https://user-images.githubusercontent.com/13024177/201789111-87c28a20-379d-4593-8770-e1b66407122a.png)

The main question here is -> do you not like the Idea of the separator? or just the actual styling?

**One last question at the end from me**

This PR expects that the first button would be shown as primary. Do we want to change it to show all buttons in this row as secondary (reffering to the conversation in the parent issue) ? As there already will be a primary button attached in the Reply area.

alexr00 reviewed (changes requested):
> The main question here is -> do you not like the Idea of the separator? or just the actual styling?

It's the syling I don't like, but we can figure that out outside of this PR. 

> Do we want to change it to show all buttons in this row as secondary (reffering to the conversation in the parent issue) ?

Yes, let's make all the additional actions buttons secondary.

Thank you for all the other changes, they look good! We'll do a UX review this week and once that (and the secondary button change is done) we should be good to merge this.

marrej commented:
Sounds good :) 
Added the "only secondary buttons" option.

**Also noticed** that there are some issues with the styling of the dropdown buttons.
![image](https://user-images.githubusercontent.com/13024177/202185147-0907c57e-5b90-4fdd-87b4-1ff455b6b0b9.png)

![image](https://user-images.githubusercontent.com/13024177/202187152-dded1d05-9e31-4a39-9c3d-1b0199b39d67.png)


so I added some changes in the `button.css` & `button.ts` to take care of the backgroundColor on separator and the whitespace caused by the border-radius
![image](https://user-images.githubusercontent.com/13024177/202185548-612ad623-df8c-44e3-9da0-c7af07450792.png)

Please let me know if you would like to keep this change or whether it should be moved to a different PR, or ignored completely.

alexr00 commented on src/vs/base/browser/ui/button/button.css:
```diff
@@ -69,6 +69,11 @@
 
 .monaco-button-dropdown > .monaco-button.monaco-dropdown-button {
 	border-left-width: 0 !important;
+	border-radius: 0 2px 2px 0;
```
@lszomoru, I think with the changes in this file you can get rid of the border-radius in https://github.com/marrej/vscode/blob/f7e26dffdc2f78e8fb75610e5950791016ffcf74/src/vs/workbench/contrib/scm/browser/media/scm.css#L290-L293 and also https://github.com/marrej/vscode/blob/f7e26dffdc2f78e8fb75610e5950791016ffcf74/src/vs/workbench/contrib/scm/browser/media/scm.css#L295-L296

alexr00 reviewed (approved):
Looks good, thanks for the PR!

alexr00 commented:
@marrej we reviewed the additional actions UX yesterday and will not be supporting the split button. Please see https://github.com/microsoft/vscode/issues/163281#issuecomment-1318401202 for the whole context.

I'll merge this PR then remove the split button components.
